core/auth/src/main/java/software/amazon/awssdk/auth/token/credentials/ProfileTokenProvider.java

```diff
@@ -37,57 +38,35 @@
  */
 @SdkPublicApi
 public final class ProfileTokenProvider implements SdkTokenProvider, SdkAutoCloseable {
-    private final SdkTokenProvider tokenProvider;
+    private SdkTokenProvider tokenProvider;
     private final RuntimeException loadException;
-
-    private final ProfileFile profileFile;
+    private final Supplier<ProfileFile> profileFile;
+    private volatile ProfileFile currentProfileFile;
     private final String profileName;
 
     /**
      * @see #builder()
      */
     private ProfileTokenProvider(BuilderImpl builder) {
-        SdkTokenProvider tokenProvider = null;
-        RuntimeException loadException = null;
-        ProfileFile profileFile = null;
-        String profileName = null;
+        RuntimeException thrownException = null;
+        Supplier<ProfileFile> selectedProfileFileSupplier = null;
+        String selectedProfileName = null;
 
         try {
-            profileName = builder.profileName != null ? builder.profileName
-                                                      : ProfileFileSystemSetting.AWS_PROFILE.getStringValueOrThrow();
-
-            // Load the profiles file
-            profileFile = Optional.ofNullable(builder.profileFile)
-                                  .orElse(builder.defaultProfileFileLoader)
-                                  .get();
-
-            // Load the profile and token provider
-            String finalProfileName = profileName;
-            ProfileFile finalProfileFile = profileFile;
-            tokenProvider =
-                    profileFile.profile(profileName)
-                               .flatMap(p -> new ProfileTokenProviderLoader(finalProfileFile, p).tokenProvider())
-                               .orElseThrow(() -> {
-                                   String errorMessage = String.format("Profile file contained no information for " +
-                                                                       "profile '%s': %s", finalProfileName, finalProfileFile);
-                                   return SdkClientException.builder().message(errorMessage).build();
-                               });
+            selectedProfileName = Optional.ofNullable(builder.profileName)
```

Comment:
NIT: I don't know what you prefer, but there is `Validate.getOrDefault(T param, Supplier<T> defaultValue)` in the `utils` module that does exactly the same logic without involving `Optional`.

Reply:
I'll leave it as is this time, but thanks for pointing out `getOrDefault` as I was not aware.